Mathlib/Analysis/Calculus/ContDiff/Basic.lean

```diff
@@ -1398,6 +1421,25 @@ theorem ContDiff.sum {ι : Type*} {f : ι → E → F} {s : Finset ι}
   simp only [← contDiffOn_univ] at *; exact ContDiffOn.sum h
 #align cont_diff.sum ContDiff.sum
 
+theorem iteratedFDerivWithin_sum_apply {ι : Type*} {f : ι → E → F} {u : Finset ι} {i : ℕ} {x : E}
```

Comment:
Perhaps it would make sense to add `[DecidableEq ι]` here to avoid using `Classical`?

Reply:
Current guideline is: if you need `DecidableEq` to **formulate** a theorem, then assume it. If only to prove it, then don't.

Reply:
Thanks! Here, we would need `DecidableEq ι` to formulate the theorem (for e.g. `Finset ι`), but we don't need it because there is an `open Classical` at the top of the file. So I guess I can leave it as is.

Reply:
`Finset ι` does not require `DecidableEq ι`. The induction principle you use does. E.g., `Finset.cons_induction_on` doesn't.

If there is a `Decidable` instance somewhere in the statement (not the proof!), then we explicitly assume it even if the file has `open Classical`. This way you don't get into trouble when you apply your lemma to `ι = Nat` or `ι = Fin n`, where you have non-classical `DecidableEq` instances.

Reply:
Thanks for the explanation! Sorry, my mistake about `Finset ι`, I had introduced another error and got confused. I see now, the statement doesn't need `DecidableEq` but the proof does due to my use of `Finset.induction`.

Reply:
Is it slightly preferable if I modify the proofs to use `Finset.cons_induction` instead?

It seems like `contDiff_update` and `contDiff_single` is this same file don't follow the principle of explicitly assuming `Decidable` even where the file has `open Classical`? (Both statements require `DecidableEq`, for `Function.update` and `Pi.single` respectively)

Reply:
OTOH, all our topology depends on `Classical.choice`, so it doesn't matter too much. One bonus of `cons_induction` is that you have `a ∉ s` as an assumption, so lemmas like `Finset.sum_cons` don't need you to repeat it.

Reply:
OK, great! I'll updated the `sum` theorem to use `Finset.cons_induction` (still with no assumption of `DecidableEq`, and it won't break if `open Classical` is removed). I've kept `Finset.induction` for the `prod` theorem as the statement requires `DecidableEq` anyway (e.g. for `Finset.sym`, `Multiset.multinomial`) and the proof then uses the `insert` form for `Finset.symInsertEquiv`. Thanks for the guidance!

I'll do a squash + rebase as this is a few hundred commits behind master.